algebird-core/src/main/scala/com/twitter/algebird/maps/prefixsum.scala

```diff
@@ -0,0 +1,224 @@
+ * @tparam K The key type
+ * @tparam V The value type
+ * @tparam P The prefix sum type
+ * @tparam IN The node type of the concrete internal R/B tree subclass
+ * @tparam M The map self-type of the concrete map subclass
+ */
+trait PrefixSumMapLike[K, V, P, IN <: INodePS[K, V, P], M <: PrefixSumMapLike[K, V, P, IN, M] with SortedMap[K, V]]
+  extends NodePS[K, V, P] with OrderedMapLike[K, V, IN, M] {
+
+  /**
+   * A container of all prefix sums over the stored values.  If 'open' is true,
+   * the sums will be for strictly < each key.
+   */
+  def prefixSums(open: Boolean = false) = prefixSumsIterator(open).toIterable
+
+  /**
+   * Iterate over prefix sums for stored values.  If 'open' is true,
+   * the sums will be for strictly < each key.
+   */
+  def prefixSumsIterator(open: Boolean = false) = {
+    val itr = valuesIterator.scanLeft(prefixMonoid.zero)((p, e) => prefixMonoid.inc(p, e))
+    if (open) itr.takeWhile(_ => itr.hasNext) else itr.drop(1)
+  }
+
+  /** equivalent to prefixSum of the right-most key */
+  def sum = this match {
+    case n: INodePS[K, V, P] => n.prefix
+    case _ => prefixMonoid.zero
+  }
+}
+
+sealed trait PrefixSumMap[K, V, P] extends SortedMap[K, V]
+  with PrefixSumMapLike[K, V, P, INodePS[K, V, P], PrefixSumMap[K, V, P]] {
+
+  override def empty = PrefixSumMap.key(keyOrdering).value[V].prefix(prefixMonoid)
+
+  override def toString =
+    "PrefixSumMap(" +
+      iterator.zip(prefixSumsIterator())
+      .map(x => s"${x._1._1} -> (${x._1._2}, ${x._2})").mkString(", ") +
+      ")"
+}
+
+object PrefixSumMap {
+  /**
+   * Instantiate a new empty PrefixSumMap from key, value and prefix types
+   * {{{
+   * import com.twitter.algebird.maps.prefixsum._
+   *
+   * // map strings to integers, using default ordering and standard integer monoid
+   * val map1 = PrefixSumMap.key[String].value[Int].prefix(IncrementingMonoid.fromMonoid[Int])
+   * // Use a custom ordering
+   * val ord: Ordering[String] = ...
+   * val map2 = PrefixSumMap.key(ord).value[Int].prefix(IncrementingMonoid.fromMonoid[Int])
+   * }}}
+   */
+  def key[K](implicit ord: Ordering[K]) = infra.GetValue(ord)
+
+  object infra {
+    /** Mediating class between key method and value method */
+    case class GetValue[K](ord: Ordering[K]) {
+      def value[V] = GetPrefix[K, V](ord)
+    }
+
+    /** Mediating class between value method and prefix method */
+    case class GetPrefix[K, V](ord: Ordering[K]) {
+      def prefix[P](implicit mon: IncrementingMonoid[P, V]): PrefixSumMap[K, V, P] =
+        new Inject[K, V, P](ord, mon) with LNodePS[K, V, P] with PrefixSumMap[K, V, P]
+    }
+  }
+}
+
+/**
+ * A monoid that also supports an 'increment' operation.  This class is intended to encode
+ * semantics similar to Scala's seq.aggregate(z)(seqop,combop), where the standard
+ * Monoid 'plus' corresponds to 'combop' and the 'inc' method corresponds to 'seqop'.
+ */
+trait IncrementingMonoid[T, E] extends Monoid[T] {
```

Comment:
After working through #501, I think I can replace my `IncrementingMonoid` with `MonoidAggregator`, since `append` is the analog of `inc`